torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -213,67 +213,22 @@ class ProcessGroupNCCL : public ProcessGroup {
         at::IValue value,
         std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents)
         : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          value_(std::move(value)),
           cudaEvents_(std::move(cudaEvents)) {
+      markCompleted(std::move(value));
     }
 
-    FutureNCCL(at::TypePtr type) : at::ivalue::Future(std::move(type)) {}
+    using at::ivalue::Future::Future;
 
-    // Gets the current stream of the device and synchronizes recorded streams
-    // with that. It will return after synchronizing the correct GPU streams to
-    // ensure we can have async CUDA execution and it does not wait for the
-    // entire operation to complete on GPU.
-    void wait() override {
-      if (error_) {
-        throw *error_;
+    void setDataPtrExtractor(DataPtrExtractor data_ptr_extractor) override {
+      // To avoid races with other threads that may be using the extractor, we
```

Comment:
do we expect this function to be called multiple times? If no, do we need a lock + assert?

Reply:
Yes, indeed I was being sloppy here, thanks for calling me out. It can be called multiple times: once when the future is constructed (if it's a "child" future created by the `then()` method), and then one more time by PythonFutureWrapper for each time we call `then()` on that future (in order to create a "grand-child" future). Now that I think of it this data race was probably there before, but we should still fix it.

I'm still hoping we can come up with a better idea for this whole DataPtrExtractor, or get rid of it entirely if we can get a JIT helper that can do it for us. If neither of that happens, I'll add a lock here.